### ISSUE TITLE
3.3v?

### ISSUE THREAD

**Issue by arpruss** [Dhole/gb-rw-stm32f411#1]
Have you had any issues due to the stm32 using 3.3v and the GameBoy cartridges expecting 5v? Do you set the VCC for the GameBoy cart to 5v or 3.3v?

**Comment by Dhole** [comment]
On the signal level, I haven't had an issue. For input signals, the STM32F4 has many pins which are 5V tolerant.  For output signals, the gameboy bus uses TTL, which will read 3.3V as 1 (source: http://friedtj.free.fr/gb_eng.pdf about TTL levels: https://learn.sparkfun.com/tutorials/logic-levels/all).

I do connect the VCC of the cart to 5V when it's an original gameboy / gameboy color cart.  For Chinese cartridge clones I use 3.3V (because internally they use 3.3V flash memories), and for GBA carts I also use 3.3V.  In the PCB I designed I included a switch to allow powering the cart at 3.3V or 5V: https://github.com/Dhole/gb-cart-nucleo-shield#pictures

Since I'm powering the STM32F4 via USB, I have easy access to 5V, which I use.

**Comment by arpruss** [comment]
Thanks! I'm trying to read a GB Tetris PCB I found on the sidewalk near my house, and have been getting inconsistent results (e.g., starting around address 0x0700, repeating reads give different results each time), and I thought it might be due to the fact that I was using a 3.3V device (stm32f1 in my case; I am using my own code) for setting the address lines. It's good to know about the GameBoy using TTL levels, as that eliminates one hypothesis about what's wrong.

**Comment by arpruss** [comment]
I am having a very odd thing happen. I power the cart from USB, and when I set the eight data lines to input on the stm, I get complete junk. But when I set the data lines to input with pulldown, then I get data that is generally right, but with occasional errors (multiple reads give different values!). I understand if you don't have time to help random strangers, but any hints would be appreciated.

**Comment by Dhole** [comment]
> I am having a very odd thing happen. I power the cart from USB, and when I set the eight data lines to input on the stm, I get complete junk. But when I set the data lines to input with pulldown, then I get data that is generally right, but with occasional errors (multiple reads give different values!). I understand if you don't have time to help random strangers, but any hints would be appreciated.

Let's see.  You should always set the GPIOs as input + pulldown for reading signal with the gameboy.  It's normal that you get garbage if the inputs are floating, because the cart probably doesn't have any pull-up / pull-down resistor internally for output signals.  Probably for output the cart sets the signal to 5V if 1, or nothing (floating) if 0; so by pulling down is when you get 0V for 0.

The fact that you manage to get data generally right is a good signal!  Something that you should look into is timings.  I remember when I started this project I had similar issues to the ones you describe.  In the beginning I would get random errors, and here are two things I have observed:
1. Timing is quite critical, specially if you want to read fast.  I understand that you're using a different microcontroller which may run at a different frequency.  If you take a look for example at this function: https://github.com/Dhole/gb-rw-stm32f411/blob/de51d105ddd2956be9f18d31e58f68a4cfa9bcef/src/gb-rw.c#L499 you'll see the usage of the `NOP_REP` macro to wait some time.  I iterated those values through many experiments until they worked.  You may need to try playing with those values until it works for your microcontroller.
2. Cartridge contact.  Sometimes the cart pins can make bad contact with the cart slot.  If you encounter errors try taking the cart out and inserting it again, or try with another cart.

Also, sometimes the errors will have patters, for example if one data pin in particular is not being red well.  I suggest that after several readings where some bytes differ, you analyze those bytes in binary, to see if the bit flips happen always at the same position.  Also try to compare the bad bytes with the original ROM, and try to find patterns there.  Some microcontrollers boards have some GPIO designed for special purposes and may give weird results when they are used as input or output.  I remember this happening to me (I don't remember which microcontroller exactly), where one bit of data was giving me many errors and I found out that the PIN was connected to an external oscillator or something along those lines!  So just review the datasheet of the microcontroller and the board to check any unexpected information about the GPIO pins you're using.

**Comment by arpruss** [comment]
Thanks! Good to know that my intuition about pulldown was correct. Oddly, I've looked at various Arduino GB cart reader sources online, and I don't think I saw them setting pulldown.

As for timing, two questions: (1) Is it crucial to pulse the read line like you do? Your read timings trace (https://dhole.github.io/post/gameboy_cartridge_emu_1/) shows the read line being low all the time. (2) Can I avoid the subtleties of the timing by just waiting a long time (like half a second) after setting the address?

The error seems to typically consist in a 1 bit getting transmitted as a zero. I'll have to check to see if there is anything special about these particular GPIOs. I've tried the trick of reading multiple times and OR'ing the results, but the ROM checksum is still wrong.

The errors seem to have to do with particular read addresses. I have no problem with the cart header, for instance. On the other hand, 0xFFF5 is usually really bad: repeated reads give different data, ranging from 0 to the correct value. It's kind of what you'd expect if you were sampling a floating bit, but I do have the pulldown set (and it works when I am reading from other addresses). I will have to see if what it looks like under an oscilloscope.

It could be my connections, of course. This is not like an important project that I want to put actual money into: it's just a fun challenge of "can I extract the ROM of this PCB I literally found on the sidewalk with the tools I have at home". So I don't have a female cartridge connector, and I am connecting via jumper wires stuffed into through-holes or using test clips.

**Comment by arpruss** [comment]
I put it on a scope. With the RD line grounded constantly, when I look at the data at 0xFFF5, I mostly get 00, but occasionally other values, including the correct value. The correct bit decays, getting pretty low after about 1 microsecond. Sometimes then it comes back up to full value, sometimes not. Here's a trace: https://pruss.mobi/dl/trace.jpg

With the RD line pulsed, and triggering the scope on the RD line going low, I sometimes get bits that decay over about a microsecond, and sometimes I get just zeroes.

One thing that sometimes happens is that on the laptop I see a bunch of zeroes being captured by the microcontroller, and then when I touch an oscilloscope probe (or a voltmeter probe, for that matter) to one of the data pins on the cartridge, suddenly non-zero data (with individual bits oscillating between 0 and the correct values) starts showing up.

**Comment by arpruss** [comment]
Got it! I suspect it is a damaged ROM chip. 

I ran the following procedure for a capture. For each byte, I asserted RD, delayed 1 microsecond, then read a sequence of 40-120 bytes from the data lines, de-asserted RD, paused 100 microseconds, and repeated this process for a total of about 10-20 times for the same byte. I then OR'ed the resulting data as my final byte value, and went on to the next byte. The resulting ROM loaded correctly into an emulator but the checksum was wrong. 

I then repeated the procedure for a total of 7 captures, varying some of the parameters as well as the VCC (I started with USB plus a voltage-drop diode for 4.5V, and then eventually went for full 5V). No capture had the correct checksum, but a number of them loaded correctly into the emulator. Finally, when I OR'ed the data across the 7 captures, the result had the correct ROM checksum, as well as the CRC32 matching the web data for "Tetris (World) (Rev 1)". Yay!

**Comment by arpruss** [comment]
Thanks for all the help!

What I did in my last comment was way too complicated! It turns out that something was wrong with my connection between A15 and the microcontroller (bad breadboard connection?), or else the GPIO I used on the micro to drive A15 was faulty and made the pin float. 

When I simply grounded A15, everything worked in the simplest way possible: I can just ground RD permanently, and read each byte a microsecond after setting A0-A14. No pulsing, no repetitions, no errors.

Supply voltage is irrelevant: 3.3V, 4.5V and 5V all work without any evident difference. 
Pulldown on the data pins makes no difference either.

https://gist.github.com/arpruss/59c943709e84d3b7dd4ed7faa187f2f5
